### PR TITLE
[Feature] Hardware Components Grouping

### PR DESCRIPTION
**Pull request by saikishor**
This PR allows to have hardware component grouping to allow the components to propagate errors between each other. This is very helpful if the hardware components of one of the actuator of a manipulator fails, we can propagate the error to the rest of the actuators by grouping all their hardware components to one group. In this way, we could avoid continue to execute the motion 

Closes https://github.com/ros-controls/ros2_control/issues/1457

### CODE REVIEW

**Comment by saikishor** [comment]
The failing tests are due to the Flake8 issue :)

**Review by christophfroehlich — APPROVED** [review]
LGTM, and thanks for updating the docs as well :grin:

**Review by christophfroehlich — DISMISSED** [review]

File: hardware_interface/doc/hardware_interface_types_userdoc.rst [hunk 1]
```diff
@@ -41,6 +41,12 @@ The ``<gpio>`` tag can be used as a child of all three types of hardware compone
...
 
+Hardware Groups
+*****************************
+Hardware Component Groups serve as a critical organizational mechanism within complex systems, facilitating error handling and fault tolerance. By grouping related hardware components together, such as actuators within a manipulator, users can establish a unified framework for error detection and response.
+
+Hardware Component Groups play a vital role in propagating errors across interconnected hardware components. For instance, in a manipulator system, grouping actuators together allows for error propagation. If one actuator fails within the group, the error can propagate to the other actuators, signaling a potential issue across the system. By default, the actuator errors are isolated to their own hardware component, allowing the rest to continue operation unaffected. In the provided ROS2 control configuration, the ``<group>`` tag within each ``<ros2_control>`` block signifies the grouping of hardware components, enabling error propagation mechanisms within the system.
```
Comment (bmagyar) [inline hardware_interface/doc/hardware_interface_types_userdoc.rst:48]:
```suggestion
Hardware Component Groups play a vital role in propagating errors across interconnected hardware components. For instance, in a manipulator system, grouping actuators together allows for error propagation. If one actuator fails within the group, the error can propagate to the other actuators, signaling a potential issue across the system. By default, the actuator errors are isolated to their own hardware component, allowing the rest to continue operation unaffected. In the provided ros2_control configuration, the ``<group>`` tag within each ``<ros2_control>`` block signifies the grouping of hardware components, enabling error propagation mechanisms within the system.
```

File: hardware_interface/src/resource_manager.cpp [hunk 2]
```diff
@@ -884,6 +898,27 @@ class ResourceStorage
...
+    const std::string & group_name, const return_type & value)
+  {
+    // This is for the components that has no configured group
+    if (group_name.empty())
+    {
+      return value;
+    }
+    // If it is anything other than OK, change the return type of the hardware group state
+    // to the respective return type
+    if (value > hw_group_state_.at(group_name))
```
Comment (bmagyar) [inline hardware_interface/src/resource_manager.cpp:915]:
```suggestion
    if (value != return_type::OK)
```
the docs clearly say what you wanted to write in code here ;) 
>  // If it is anything other than OK

**Review by bmagyar — CHANGES_REQUESTED** [review]

**Review by bmagyar — APPROVED** [review]

**Pull request merged**